### ISSUE TITLE
Unlock Protocol-powered, members only Farcaster Channels

### ISSUE THREAD

**Issue by julien51** [unlock-protocol/unlock#13890]
The goal here is to create an application that would integrate _completely_ in the Farcaster ecosystem, thru both channels and frames (and maybe actions).

Let's start with user stories:

- As a "creator" I have already set-up a farcaster channel and I want to configure it so that only my fans (members) can join. I want to choose whether I allow them to "read" or "write" messages (and maybe, later, allow them to do either based on a different membership level).
- As a "fan", I want to join the channels of my favorite creators and read their casts (or write depending on permissions set by the creator).

@nestorbonilla

**Comment by julien51** [comment]
Assumptions (which can be revised later):
- creators have already deployed their channel
- creators will add the Unlock token-gating mechanism through the "moderation" feature
- fans will use frames to "purchase/mint" their NFT memberships (possible to use checkout URL for debugging purposes initially)

**Comment by nestorbonilla** [comment]
Hi @julien51, here's a breakdown of the plan based on our chat and the research I've done:

**Custom Hook Implementation:**
Create custom hooks to trigger actions when a cast is submitted to a specific channel.
_Hook Trigger:_ New cast detected on the designated channel.

**Data Extraction:**

- Retrieve the FID (Farcaster ID) associated with the cast.
- Retrieve the Ethereum address(es) associated with the FID that created the cast.
- Query the Unlock Protocol's ERC721 contract to:
  - Determine the balance of Unlock NFTs held by the address(es).
  - Validate any additional conditions specified in the NFT contract (e.g., subscription expiration date, membership tier, referral, etc).

**Validation & Action:**

- Scenario 1 (User Owns Unlock NFT):
  - Upvote the cast to show in the feed.
- Scenario 2 (User Doesn't Own Unlock NFT):
  - Initiate a Direct Cast (DC) to the user.
  - Include a link in the DC suggesting the purchase of an Unlock NFT.

**Notes:**

- Farcaster's current design doesn't allow for preventing casts. The primary focus is on moderating the feed by upvoting or downvoting content.
- Currently, DC doesn't allow frame rendering, but this feature is [coming soon](https://warpcast.com/horsefacts.eth/0x2bed1aa6), so I'll build the application with that in mind to enable a smoother purchase flow in the future.

**Comment by julien51** [comment]
> Currently, DC doesn't allow frame rendering, but this feature is [coming soon](https://warpcast.com/horsefacts.eth/0x2bed1aa6), so I'll build the application with that in mind to enable a smoother purchase flow in the future.

Instead of a DM, can we just send an @reply to the user (outside of any channel) ? This way that would render as a frame...

**Comment by nestorbonilla** [comment]
Yes, we can. To confirm, you'd like the bot to create a public cast (not on the channel) that tags the user, including a frame to do the purchase flow, instead of sending a DM. Is that right?

**Comment by julien51** [comment]
@nestorbonilla please post updates here :)

**Comment by nestorbonilla** [comment]
Great, sharing some updates:

**Custom Hooks**: Implementation complete for triggering actions on new casts to a specific channel.
**Next Up**: Scalability. I'm opting to use a hook per channel for now, while still exploring options for more dynamic multi-channel management in the future.
#### Validation & Action Refinements:
- **Scenario 1 (NFT Owner)**: Cast upvoted. 👍
- **Scenario 2 (No NFT)**: Public @reply to the user with a mint/purchase frame. I'll explore the DC option too.
- **Scenario 3 (Channel Owner)**: Initial implementation using a moderation frame triggered by `@botname setup`. The bot will reply with an embedded frame for chain, contract, and referral selection to gate main feed access.

**Comment by nestorbonilla** [comment]
Sharing some updates:

- Scenarios 1 (NFT owner casting) and 3 (channel owner setup) are implemented.
- Channel leads can now add or remove Unlock token gating for their channels directly within the moderation frame.
- Implemented frame validations to ensure only authorized users can modify channel settings.

Next Steps:

- Develop a purchase flow frame for new memberships.
- Develop renewal flow frame for expired memberships.

**Comment by nestorbonilla** [comment]
Hi @julien51, I've completed the implementation of all planned features for the bot:

**Here's a quick recap of what's been done:**

- **Token-Gating:** Channel owners can now easily add/remove token requirements using the `@membersonly setup` command, which triggers a configuration frame for setting the desired membership contract.
- **Purchase Flow:** A streamlined purchase flow has been implemented where non-members attempting to cast on a gated channel are automatically tagged in a new cast with an embedded frame, allowing them to easily purchase or renew their membership.
- **Membership Hooks:** The application validates membership status for each cast, ensuring that only valid members' content is promoted to the main feed.

The code is publicly available [here](https://github.com/nestorbonilla/members-only). Let me know if you have any feedback!
